### ISSUE TITLE
"unauthorized" on opensubtitles

### ISSUE THREAD

**Issue by zandadoum** [morpheus65535/bazarr#881]
good morning.
i seem to have a little problem:

> Throttling opensubtitles for 10 minutes, until 20/03/24 09:19, because of: Unauthorized. Exception info: None

the error is "Unauthorized" all the time.
I tried with linuxserver-bazarr:latest and linuxserver-bazarr:development with same problem
I also made sure my password work, i went to the opensubtitles web and could log in correctly and I have no limitations or anything that could prevent me from getting subs
also tried without a username/password and disabling SSL, same error
this has been going on for a few days now

**Comment by lukjod** [comment]
+1

Bazarr Version:
0.8.4.2
Sonarr Version:
2.0.0.5344
Radarr Version:
0.2.0.1480
Operating System:
Linux-4.14.24-qnap-x86_64-with
Python Version:
3.8.2
Bazarr Directory:
/app/bazarr
Bazarr Config Directory:
/config

log:
Using opensubtitles again after 10 minutes, (disabled because: Unauthorized)

**Comment by BodingClockchian** [comment]
+1
Bazarr Version:
0.8.4.2
Sonarr Version:
3.0.3.735
Radarr Version:
0.2.0.1480
Operating System:
Linux-5.3.0-42-generic-x86_64-with 
Python Version:
3.8.2

Followed by logging :

> 
	Throttling opensubtitles for 10 minutes, until 20/03/24 11:26, because of: Unauthorized. Exception info: None	
a few seconds ago
	Unexpected error in provider 'opensubtitles': Traceback (most recent call last): File "/app/bazarr/bazarr/../libs/subliminal_patch/core.py", line 168, in list_subtitles_provider results = self[provider].list_subtitles(video, provider_languages) File "/app/bazarr/bazarr/../libs/subliminal_patch/providers/opensubtitles.py", line 237, in list_subtitles return self.query(languages, hash=video.hashes.get('opensubtitles'), size=video.size, imdb_id=video.imdb_id, File "/app/bazarr/bazarr/../libs/subliminal_patch/providers/opensubtitles.py", line 270, in query response = self.use_token_or_login( File "/app/bazarr/bazarr/../libs/subliminal_patch/providers/opensubtitles.py", line 161, in use_token_or_login self.log_in() File "/app/bazarr/bazarr/../libs/subliminal_patch/providers/opensubtitles.py", line 147, in log_in response = self.retry( File "/app/bazarr/bazarr/../libs/subliminal_patch/providers/mixins.py", line 47, in retry return f() File "/app/bazarr/bazarr/../libs/subliminal_patch/providers/opensubtitles.py", line 148, in lambda: checked( File "/app/bazarr/bazarr/../libs/subliminal_patch/providers/opensubtitles.py", line 372, in checked raise Unauthorizedsubliminal.providers.opensubtitles.Unauthorized	
a few seconds ago
	https://api.opensubtitles.org:443 "POST /xml-rpc HTTP/1.1" 200 177	
a few seconds ago
	No subtitles found	
a few seconds ago
	https://www.podnapisi.net:443 "GET /subtitles/search/old?sXML=1&sL=nl&sK=miss+fishers+murder+mysteries&sTS=2&sTE=2&sY=2012 HTTP/1.1" 200 1465	
a few seconds ago
	DNS: Falling back to default DNS or IP on api.opensubtitles.org	
a few seconds ago
	Starting new HTTPS connection (1): api.opensubtitles.org:443	
a few seconds ago
	Searching subtitles [{'moviehash': '1be0110efb9f9424', 'moviebytesize': '2326697573', 'sublanguageid': 'dut'}, {'query': 'XXXXXX', 'season': 2, 'episode': 2, 'sublanguageid': 'dut'}]	
a few seconds ago
	Login failed, please check your credentials	
a few seconds ago
	https://api.opensubtitles.org:443 "POST /xml-rpc HTTP/1.1" 200 177
>

**Comment by GermanG** [comment]
Some debugging:

Does not look like an API issue, short python to test Login

```
from urllib3.packages.six.moves.xmlrpc_client import ServerProxy

server = ServerProxy('https://api.opensubtitles.org/xml-rpc')
response = server.LogIn('############', '##########', 'eng', 'Bazarr/1')

print(response)
```

Somethig odd I've found in the logs:
```
2020-03-24 17:02:46,451 - subliminal_patch.providers.opensubtitles (7f8e0cf11700) :  DEBUG (opensubtitles:184) - Trying previous token: '<dogpile.cXXXXXXXXXXXXXXXXXXXXXXXX'
```
This does not look like a token, anyway it tries to use my credentials and fails.

**Comment by morpheus65535** [comment]
Thanks @GermanG for looking at it. If you are able to go further in your investigation, it would be greatly appreciated! I'm a bit overwhelmed by work with the actual crisis. :-/

**Comment by svenvg93** [comment]
Some problem, password is working on the website.

some debug loggin;
`24/03/2020 19:18:44|DEBUG   |root                            |BAZARR is using these video object properties: {'name': '/tv/Chicago Fire/Season 08/Chicago.Fire.S08E17.720p.AMZN.WEBRip.DDP5.1.x264-KiNGS[rarbg].mkv', 'format': 'WEBRip', 'release_group': 'KiNGS', 'resolution': '720p', 'video_codec': 'h264', 'audio_codec': 'EAC3', 'imdb_id': None, 'hashes': {'opensubtitles': 'e25e920ecb1eed10'}, 'size': 1743184175, 'subtitle_languages': set(), 'original_name': 'Chicago.Fire.S08E17.Protect.a.Child.720p.AMZN.WEB-DL.DDP5.1.H.264-KiNGS.mkv', 'plexapi_metadata': {}, 'hints': {'title': 'Chicago Fire', 'type': 'episode', 'single_value': True, 'expected_title': ['Chicago Fire']}, 'audio_languages': set(), 'external_subtitle_languages': set(), 'series': 'Chicago Fire', 'season': 8, 'episode': 17, 'title': 'Protect a Child', 'year': 2012, 'original_series': True, 'tvdb_id': None, 'series_tvdb_id': 258541, 'series_imdb_id': None, 'alternative_series': [], 'used_scene_name': True, 'original_path': '/tv/Chicago Fire/Season 08/Chicago.Fire.S08E17.Protect.a.Child.720p.AMZN.WEB-DL.DDP5.1.H.264-KiNGS.mkv', 'fps': 23.976}|
24/03/2020 19:18:44|INFO    |subliminal_patch.core           |Listing subtitles for <Episode ['Chicago Fire', 2012, 8x17]>|
24/03/2020 19:18:44|INFO    |subliminal_patch.core           |Listing subtitles with provider 'opensubtitles' and languages {<Language [nl]>}|
24/03/2020 19:18:44|INFO    |subliminal_patch.core           |Initializing provider opensubtitles|
24/03/2020 19:18:44|DEBUG   |subliminal_patch.providers.opensubtitles|Using timeout: 15|
24/03/2020 19:18:44|DEBUG   |subliminal_patch.providers.opensubtitles|Using HTTPS connection|
24/03/2020 19:18:44|INFO    |subliminal_patch.providers.opensubtitles|Logging in|
24/03/2020 19:18:44|DEBUG   |subliminal_patch.providers.opensubtitles|Trying previous token: 'bytearray(XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX'|
24/03/2020 19:18:44|DEBUG   |urllib3.connectionpool          |Starting new HTTPS connection (1): api.opensubtitles.org:443|
24/03/2020 19:18:44|DEBUG   |subliminal_patch.http           |DNS: Falling back to default DNS or IP on api.opensubtitles.org|
24/03/2020 19:18:44|DEBUG   |urllib3.connectionpool          |https://api.opensubtitles.org:443 "POST /xml-rpc HTTP/1.1" 200 212|
24/03/2020 19:18:44|DEBUG   |subliminal_patch.providers.opensubtitles|Token not valid.|
24/03/2020 19:18:44|DEBUG   |urllib3.connectionpool          |Starting new HTTPS connection (1): api.opensubtitles.org:443|
24/03/2020 19:18:44|DEBUG   |subliminal_patch.http           |DNS: Falling back to default DNS or IP on api.opensubtitles.org|
24/03/2020 19:18:44|DEBUG   |urllib3.connectionpool          |https://api.opensubtitles.org:443 "POST /xml-rpc HTTP/1.1" 200 177|
24/03/2020 19:18:44|ERROR   |subliminal_patch.providers.opensubtitles|Login failed, please check your credentials|
24/03/2020 19:18:44|INFO    |subliminal_patch.providers.opensubtitles|Searching subtitles [{'moviehash': 'e25e920ecb1eed10', 'moviebytesize': '1743184175', 'sublanguageid': 'dut'}, {'query': 'Chicago Fire', 'season': 8, 'episode': 17, 'sublanguageid': 'dut'}]|
24/03/2020 19:18:44|DEBUG   |urllib3.connectionpool          |Starting new HTTPS connection (1): api.opensubtitles.org:443|
24/03/2020 19:18:44|DEBUG   |subliminal_patch.http           |DNS: Falling back to default DNS or IP on api.opensubtitles.org|
24/03/2020 19:18:44|DEBUG   |urllib3.connectionpool          |https://api.opensubtitles.org:443 "POST /xml-rpc HTTP/1.1" 200 177|
24/03/2020 19:18:44|ERROR   |subliminal_patch.core           |Unexpected error in provider 'opensubtitles': Traceback (most recent call last):  File "/app/bazarr/bazarr/../libs/subliminal_patch/core.py", line 168, in list_subtitles_provider    results = self[provider].list_subtitles(video, provider_languages)  File "/app/bazarr/bazarr/../libs/subliminal_patch/providers/opensubtitles.py", line 237, in list_subtitles    return self.query(languages, hash=video.hashes.get('opensubtitles'), size=video.size, imdb_id=video.imdb_id,  File "/app/bazarr/bazarr/../libs/subliminal_patch/providers/opensubtitles.py", line 270, in query    response = self.use_token_or_login(  File "/app/bazarr/bazarr/../libs/subliminal_patch/providers/opensubtitles.py", line 161, in use_token_or_login    self.log_in()  File "/app/bazarr/bazarr/../libs/subliminal_patch/providers/opensubtitles.py", line 147, in log_in    response = self.retry(  File "/app/bazarr/bazarr/../libs/subliminal_patch/providers/mixins.py", line 47, in retry    return f()  File "/app/bazarr/bazarr/../libs/subliminal_patch/providers/opensubtitles.py", line 148, in <lambda>    lambda: checked(  File "/app/bazarr/bazarr/../libs/subliminal_patch/providers/opensubtitles.py", line 372, in checked    raise Unauthorizedsubliminal.providers.opensubtitles.Unauthorized|Traceback (most recent call last):  File "/app/bazarr/bazarr/../libs/subliminal_patch/core.py", line 168, in list_subtitles_provider    results = self[provider].list_subtitles(video, provider_languages)  File "/app/bazarr/bazarr/../libs/subliminal_patch/providers/opensubtitles.py", line 237, in list_subtitles    return self.query(languages, hash=video.hashes.get('opensubtitles'), size=video.size, imdb_id=video.imdb_id,  File "/app/bazarr/bazarr/../libs/subliminal_patch/providers/opensubtitles.py", line 270, in query    response = self.use_token_or_login(  File "/app/bazarr/bazarr/../libs/subliminal_patch/providers/opensubtitles.py", line 161, in use_token_or_login    self.log_in()  File "/app/bazarr/bazarr/../libs/subliminal_patch/providers/opensubtitles.py", line 147, in log_in    response = self.retry(  File "/app/bazarr/bazarr/../libs/subliminal_patch/providers/mixins.py", line 47, in retry    return f()  File "/app/bazarr/bazarr/../libs/subliminal_patch/providers/opensubtitles.py", line 148, in <lambda>    lambda: checked(  File "/app/bazarr/bazarr/../libs/subliminal_patch/providers/opensubtitles.py", line 372, in checked    raise Unauthorizedsubliminal.providers.opensubtitles.Unauthorized|
24/03/2020 19:18:44|INFO    |root                            |Throttling opensubtitles for 10 minutes, until 20/03/24 19:28, because of: Unauthorized. Exception info: None|
24/03/2020 19:18:44|INFO    |subliminal_patch.core           |Discarding provider opensubtitles|
24/03/2020 19:18:44|INFO    |subliminal_patch.core           |Found 0 subtitle(s)|`

**Comment by rigas40** [comment]
windows version also stop working

**Comment by GermanG** [comment]
I have a hack that "fixes" it, in libs/subliminal_patch/providers/opensubtitles.py, line 138:
```
  def get_server_proxy(self, url, timeout=None):
        return ServerProxy(url)
        # return ServerProxy(url, SubZeroRequestsTransport(use_https=self.use_ssl, timeout=timeout or self.timeout,
        #                                                 user_agent=os.environ.get("SZ_USER_AGENT", "Sub-Zero/2")))
```
Basically comment out the return ServerProxy and replace it with the one I wrote.
It's a hack and it's not a real fix, I need to understand what SubZeroRequestsTransport does.

**Comment by morpheus65535** [comment]
> It's a hack and it's not a real fix, I need to understand what SubZeroRequestsTransport does.

Thanks for taking the time to look into it. It's our most used provider with about 45% of download.

**Comment by morpheus65535** [comment]
@pannal this one should be problematic for you in sub-zero.

**Comment by GermanG** [comment]
The main issue I see so far is the lack of user-agent header in the requests (will go as 'Python-xmlrpc/3.8'), which is a no-no.

**Comment by morpheus65535** [comment]
@GermanG isn't agent set here?

https://github.com/morpheus65535/bazarr/blob/fd339b5fffa14e0e92b9d8aa7ad3de91c3e7698b/libs/subliminal_patch/providers/opensubtitles.py#L149

**Comment by GermanG** [comment]
@morpheus65535 I was referring to my "fix".
Anyway, Bazarr User-Agent is blocked. I don't know what I did wrong on my first try (changing it), but using Python original UA works:
```
        return ServerProxy(url, SubZeroRequestsTransport(use_https=self.use_ssl, timeout=timeout or self.timeout, user_agent='Python-xmlrpc/3.8'))
````
Maybe @pannal has the same issue, I tried "Sub-Zero/2" and failed too.

**Comment by morpheus65535** [comment]
@GermanG yes it's been done in January of last year. I've wrote an email to OS admin to have more informations about the current status of my user agent as there is no mention of anything on the web site.

**Comment by GermanG** [comment]
@morpheus65535 I've made a curl command to show it's the UA (replace Bazarr/1 with Python-xmlrpc/3.8):
```curl -v -X POST -H 'Accept: */*' -H 'Connection: keep-alive' -H 'Host: api.opensubtitles.org' -H 'Accept-Encoding: gzip' -H 'Content-Type: text/xml' -H 'User-Agent: Bazarr/1' --data "<?xml version='1.0'?><methodCall><methodName>LogIn</methodName><params><param><value><string>YOURUSERNAME</string></value></param><param><value><string>YOURPASSWORD</string></value></param><param><value><string>eng</string></value></param><param><value><string>Bazarr/1</string></value></param></params></methodCall>" 'http://api.opensubtitles.org/xml-rpc' --output - | gunzip - | xmllint --format -
```

On the other hand, I've read somewhere (don't remember where) that you can use the username as UA, that might be a solution too.

**Comment by morpheus65535** [comment]
@GermanG could this be a temporary fix?

https://github.com/morpheus65535/bazarr/blob/f6c303e2b576237d9908ce298429f18bc7e437d5/libs/subliminal_patch/providers/subssabbz.py#L90

**Comment by morpheus65535** [comment]
@pannal is working on a fix.

**Comment by opensubtitles** [comment]
hi all, we was forced to changed some authorization for some user agents (UAs which used totally 80% of all requests), so now during corona virus outbreak our servers was overloaded. To free up load we decided to require user authorization (username and password used on opensubtitles.org), so it will work. I did some debug (simple LogIn and SearchSubtitles on xml-rpc with Bazarr UA and it works here).

Basically you have to insert username and password into LogIn() API method and then it will works. User have to register on our site www.opensubtitles.org

**Comment by Orbeatle** [comment]
I'm still seeing an 'unauthorized' error, even though I used my correct login details in Bazarr's settings. Should I do something different?

**Comment by zandadoum** [comment]
@opensubtitles 
i was already using user/pass when i detected this bug in the first place.
i can't test right now, but unless something has changed in the past 24h. this is still not working, like @Orbeatle said.

**Comment by Levynger** [comment]
just tested it agin - using user/pass - still getting unauthorized

**Comment by morpheus65535** [comment]
@opensubtitles Thanks for taking the time to comment this issue. I agree with you that we must not try to "hack" the UA as a permanent solution and was refering to this as a temporary fix while we try to make things work like they should.

Now that we are aware of a change on your side, I can confirm that only adding username and password while calling the LogIn method doesn't change the resulting 401. You can test with curl:

```
curl -v -X POST -H 'Accept: */*' -H 'Connection: keep-alive' -H 'Host: api.opensubtitles.org' -H 'Accept-Encoding: gzip' -H 'Content-Type: text/xml' -H 'User-Agent: Bazarr/1' --data "<?xml version='1.0'?><methodCall><methodName>LogIn</methodName><params><param><value><string>USERNAME</string></value></param><param><value><string>PASSWORD</string></value></param><param><value><string>eng</string></value></param><param><value><string>Bazarr/1</string></value></param></params></methodCall>" 'http://api.opensubtitles.org/xml-rpc' --output - | gunzip - | xmllint --format -
```

Are you sure there is nothing else required?

**Comment by opensubtitles** [comment]
hi guys, there was problem with http user agent, it would work if...anyway, I fixed it now, so should work please confirm.

**Comment by morpheus65535** [comment]
@opensubtitles still getting the same result:
```
<?xml version="1.0" encoding="utf-8"?>
<methodResponse>
  <params>
    <param>
      <value>
        <struct>
          <member>
            <name>token</name>
            <value>
              <string>aaaaaaaaaaaaaaaaaaaaaaaaa</string>
            </value>
          </member>
          <member>
            <name>status</name>
            <value>
              <string>401 Unauthorized</string>
            </value>
          </member>
          <member>
            <name>seconds</name>
            <value>
              <double>0.016</double>
            </value>
          </member>
        </struct>
      </value>
    </param>
  </params>
</methodResponse>
```

**Comment by Orbeatle** [comment]
I'm now seeing another error: Throttling opensubtitles for 20 minutes, until 20/03/26 14:47, because of: ServiceUnavailable. Exception info: '503'

**Comment by BodingClockchian** [comment]
Opensubtitles is working again for me..

**Comment by morpheus65535** [comment]
Maybe sometime required to sync things on their side. Working in Bazarr now if you have a username/password set. :-)

**Comment by Orbeatle** [comment]
Yes, got it working again too. Restarted the service and reenabled opensubitles.
Cool!

**Comment by maxtheking** [comment]
Opensubtitles is working again on Bazarr.

Cheers!

**Comment by zandadoum** [comment]
@morpheus65535 @opensubtitles can confirm, with pass it's working again.
thanks for the help and solution

as a finisher, i have a unrelated side-question:
@opensubtitles i know you have a download limit/day , but don't searches also hurt your server? i do hundreds of searches a day, but only download a dozen subs.

**Comment by morpheus65535** [comment]
Thanks all for the confirmation!

**Comment by morpheus65535** [comment]
@Natejt they been hacked and you need to reset your password and change it in Bazarr.

**Comment by mbacalan** [comment]
@morpheus65535 I think .org was hacked, .com should be fine?

Either way, I had made a new account at .com (16 Feb 2022) and have always gotten the authentication error since.

@opensubtitles is the .com site also managed by you?

**Comment by morpheus65535** [comment]
@mbacalan yes .com should be fine. I was under the impression that you were talking about .org. Make sure that you can logging to os.com by copying/pasting your username nd password from config.ini file.

**Comment by opensubtitles** [comment]
@mbacalan yes, opensubtitles.com is managed by us, let us know how we can help. make sure you can login to .com

**Comment by runfabi** [comment]
This still seems to persist using the latest docker image and tried both .com as .org and the logs seems to indicate the opensubtitles request not getting authoroised.

@opensubtitles Any solution for this, as I've been through a decent amount of docker images now and I'd really like to use this again.

**Comment by GermanG** [comment]
Maybe there's a change in the way the API needs to be accessed, and change user/password to user/key.
https://www.opensubtitles.com/en/consumers
```
In order to consume our API, you first need to register a consumer here

Find our API documentation here: https://opensubtitles.stoplight.io/
```

**Comment by morpheus65535** [comment]
> Maybe there's a change in the way the API needs to be accessed, and change user/password to user/key.

We already use an consumer apikey: https://github.com/morpheus65535/bazarr/blob/master/bazarr/app/get_providers.py#L177

**Comment by morpheus65535** [comment]
@runfabi are you using your email or your username for those providers?

**Comment by rafael-maia09** [comment]
> > Maybe there's a change in the way the API needs to be accessed, and change user/password to user/key.
> 
> We already use an consumer apikey: https://github.com/morpheus65535/bazarr/blob/master/bazarr/app/get_providers.py#L177

Hi, I saw the announcement about the OpenSubtitles account at bazarr today, and I have imported my account to the .com site.
Checking the commit that you add the link, I'm seeing that's for the opensubtitles.org, not the opensubtitles.com provider.

Also, reading the https://opensubtitles.stoplight.io/docs/opensubtitles-api/e3750fd63a100-getting-started I see that we need to add a consumer api to be able to use their api, but I do not see what I need to enter in the bazarr provider configuration.
